### ISSUE TITLE
[Feature Request] Global Menu and Window Button applets

### ISSUE THREAD

**Issue by vibudh007** [pop-os/cosmic-epoch#894]
I'm being quite selfish with this feature request but essentially I would like to replicate the workflow I had with KDE a couple of years ago before updates to KDE made it impossible to achieve and I switched to GNOME which doesn't have some of these features.

Here is a screenshot of what I'm talking about
![Screenshot_20221026_173130](https://github.com/user-attachments/assets/2dc1ac04-a761-453c-aa1b-21e68b206250)

In the top panel, there are a few important extensions. Namely, the Global Menu and the Window Buttons. The Media Player is also useful but not as important.

With the Global Menu and the Window Buttons, I could delegate the tasks of the top bar of every window to the top panel of the desktop, and remove the top bar of each window altogether. This would give me more space on the screen for the windows. 
I could also conveniently pull my mouse to the top right of the screen and left click to close the active window without having to look for the close button of that window.

In the screenshot, the top bar is compact and doesn't stretch to the whole screen but I would usually keep it stretched to the whole screen.

I miss how clean and efficient this felt. I've tried cosmic and it's 90% of the way there to be able to replicate this workflow, but is missing those couple of applets at the moment.

If the cosmic devs see any potential in either any of the applet ideas or the idea of being able to turn off the top bar of windows, then I would love to see them implemented.

**Comment by Drakulix** [comment]
We get a request for a global menu a lot. The big problem here is, that there isn't a unified protocol for this to work reliably across apps. Displaying the title and window controls on the other hand would be trivial.

**Comment by vibudh007** [comment]
@Drakulix Will there be a way to disable window top bars as well?

**Comment by Drakulix** [comment]
If we would add such an applet, it should be accompanied by removing window headers, however I don't think we have concrete plans for that yet.

For that at least a protocol exists, however some apps might not respect that or decide to only hide window controls from their top bar.

**Comment by Hate9** [comment]
It's worth stating for the record that while a global menu is (sadly) not yet implemented for Cosmic, you *absolutely can* put your "window buttons" in the Panel.
If you go to Settings>Desktop>Panel>Applets, you should be able to add "App Tray", which is the same "window button" applet otherwise found in the dock at the bottom.
I, too, hate having more than one panel/dock taking up screen space.

**Comment by vibudh007** [comment]
@Hate9 By "Window Buttons", I did not mean the app tray, I meant the minimize, maximize and close buttons. Having one set of them in the panel instead of one set for each open window makes it so that there's some screen space saved (At least in Qt applications) and also you can quickly flick your mouse to the top right of the screen to close the currently focused window.

**Comment by Sporesirius** [comment]
> We get a request for a global menu a lot. The big problem here is, that there isn't a unified protocol for this to work reliably across apps...

Hi, would a Wayland protocol be a possible solution for this, which toolkits could implement and compositors could decide whether to support the feature?

**Comment by Munteanujasmin09** [comment]
What you think about kde realisation of global menu in wayland?:
https://wayland.app/protocols/kde-appmenu

**Comment by Drakulix** [comment]
> > We get a request for a global menu a lot. The big problem here is, that there isn't a unified protocol for this to work reliably across apps...
> 
> @Drakulix Hi, would a Wayland protocol be a possible solution for this, which toolkits could implement and compositors could decide whether to support the feature?

The protocol is not the problem, several attempts exist (e.g. kde-appmenu). The problem is getting toolkits to adopt it.

**Comment by Munteanujasmin09** [comment]
But i we can create custom applet that will add kde-appmenu (even if will work only with qt apps)? Can libcosmic devs add support for appmenu?

**Comment by Sporesirius** [comment]
> > > We get a request for a global menu a lot. The big problem here is, that there isn't a unified protocol for this to work reliably across apps...
> > 
> > 
> > @Drakulix Hi, would a Wayland protocol be a possible solution for this, which toolkits could implement and compositors could decide whether to support the feature?
> 
> The protocol is not the problem, several attempts exist (e.g. kde-appmenu). The problem is getting toolkits to adopt it.

Is this a case of "GTK/GNOME developers having a different vision, so they won't support the protocol—like with SSD?"

**Comment by Drakulix** [comment]
> Is this a case of "GTK/GNOME developers having a different vision, so they won't support the protocol—like with SSD?"

I wouldn't necessarily say so. Their gtk-shell protocol has an option to set a menupath: https://wayland.app/protocols/gtk-shell#gtk_surface1:request:set_dbus_properties

Which doesn't seem to be used by GTK4/Adwaita apps anymore, but I am not sure there exists a statement from gnome developers on that issue.

What this needs is a common standard, but I am not sure anybody has tried to make a wayland-protocols proposal.

(Which is not an invite to open an issue at wayland-protocols. Somebody needs to actually write the protocol (which could be based on kde-appmenu), collect feedback and then drive toolkit implementations.)

As for implementing the protocol as is in COSMIC: No, I don't want to support an unstable third-party protocol. If there is common interest, that is a good reason to start standardizing.

**Comment by hyuri** [comment]
I'd like to cast another vote of support for global menus. Beyond the perspective of customizability, to me, the biggest value in offering it is making the DE more welcoming and accessible to users coming from years of experience on macOS, building that muscle memory. So it's part of a bigger issue of offering a UX that allows users of the bigger platforms — Windows and macOS — to feel more at home.

**Comment by Sporesirius** [comment]
Hi,
It seems that there was/is work going on for a more generic protocol: https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/52

**Comment by CosmicFusion** [comment]
I would really love this, I'm using unite (window controls) + fildem (global menu) on my GNOME 48, it would be awesome to carry over my setup to cosmic

**Comment by usbeehu** [comment]
> Hi, It seems that there was/is work going on for a more generic protocol: https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/52

Will this ever happen?
There is a KDE [blog post](https://blogs.kde.org/2020/10/11/linux-desktop-shell-ipc-wayland-vs-d-bus-and-lack-agreement-when-use-them/) that had it's 5th birthday last saturday. Literally nothing happened since then.

As far as I know Wayland lets create protocols for a specific implementation/compositor, so it would be possible to add a Cosmic protocol to cosmic-comp so at least there would be a protocol to start with. Later KDE could adopt that too if they want to. But for now it seems like they are playing the waiting game rather than making their own protocol for KWin (for wayland at least).
If my understandings are correct and this is true it would be possible to create a Cosmic specific implementation of dbus_annotation, rather than waiting forever.

It seems like Gnome devs will never do something like this, Wayland devs are uninterested, so the two community, namely KDE and Cosmic are waiting on others rather than doing it themselves.

**Comment by palaashatri** [comment]
Looks like this is a low priority issue/not under consideration for right now, considering no communication from project maintainers since 2024, and new issues being marked as dupes of this one.

**Comment by zero-hugo** [comment]
I'm looking forward to it！！！

**Comment by dimparf** [comment]
Please add this functionality.

**Comment by usbeehu** [comment]
That's what I'm thinking too. If we want this to happen there should be a Cosmic protocol for this just like how there are KDE protocols for other things too.

**Comment by ratnik2** [comment]
It seems there is renewed interest in standardizing a protocol that continued off the old protocol [(old protocol: !52)](https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/52):

new protocol: https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/493

_The protocol allows surfaces to be associated_
_with DBus objects and interfaces._
_There's several use cases for this:_

_associating surfaces with (AT-SPI) accessibility trees_
**associating surfaces with DBus-exported menus**

_The protocol is designed in such a way that it allows apps/compositors_
_to use it for additional tasks without spec changes, allowing to replace_
_custom protocols currently in use for this task._

Just needs 2 more acks and a final review. This looks promising.
